File: amazon/93_restore_IP_adresses.py

class Solution:
    def restoreIpAddresses(self, s: str) -> List[str]:
        result=[]
        self.backtrack(s, 0,0, "", result)
        return result

    def backtrack(self, s, start, dots, path, result):
        if dots == 4 and start == len(s):
            result.append(path[:-1]) # remove trailling dot
            return
        if dots > 4 or start >= len(s):
            return
        for length in range(1,4):
            if start + length > len(s):
                break
            segment = s[start:start+ length]
            if self.isValid(segment):
                self.backtrack(s, start+length, dots+1, path+segment+'.', result)
            
    def isValid(self, segment):
        if len(segment) > 3 or (segment.startswith('0') and len(segment) > 1):
            return False
        val = int(segment)
        return 0 <= val <= 255



#iterative approach using nested loops
class Solution:
    def restoreIpAddresses(self, s:str) -> List[str]:
        result = []
        n = len(s)
        for i in range(1, min(n, 4)):
            for j in range(i+1, min(n, i+4)):
                for k in range(j+1, min(n, j+4)):
                    s1 = s[:i]
                    s2 = s[i:j]
                    s3 = s[j:k]
                    s4 = s[k:]
                    if all(map(self.isValid, [s1,s2,s3,s4])):
                        result.append(f"{s1}.{s2}.{s3}.{s4}")
        return result

    def isValid(self, segment):
        if len(segment) == 0 or len(segment) >3:
            return False
        if segment.startswith('0') and len(segment) >1:
            return False
        val = int(segment)
        return 0 <= val <= 255

File: amazon/93_restore_IP_adresses.cpp
// A valid IP address consists of exactly four integers separated by single dots. 
// Each integer is between 0 and 255 (inclusive) and cannot have leading zeros.

// For example, "0.1.2.201" and "192.168.1.1" are valid IP addresses, 
// but "0.011.255.245", "192.168.1.312" and "192.168@1.1" 
// are invalid IP addresses.
// Given a string s containing only digits, return all possible valid IP 
// addresses that can be formed by inserting dots into s. 
// You are not allowed to reorder or remove any digits in s. 
// You may return the valid IP addresses in any order.

 

// Example 1:
// Input: s = "25525511135"
// Output: ["255.255.11.135","255.255.111.35"]

// Example 2:
// Input: s = "0000"
// Output: ["0.0.0.0"]

// Example 3:
// Input: s = "101023"
// Output: ["1.0.10.23","1.0.102.3","10.1.0.23","10.10.2.3","101.0.2.3"]
 

// Constraints:
// 1 <= s.length <= 20
// s consists of digits only.



//recursive backtracking
class Solution{
public: 
    vector<string> restoreIpAddresses(string s){
        vector<string> result;
        backtrack(s,0,0,"",result);
        return result;
    }
private:
    void backtrack(const string &s, int start, int dots, string path, vector<string> &result){
        if(dots == 4 && start == s.size()){
            result.push_back(path.substr(0, path.size()-1)); // remove trailling dots
            return;
        }
        if(dots > 4 || start >= s.size()){
            return;
        }
        for(int len = 1; len <= 3 && start + len <= s.size(); len++){
            string segment = s.substr(start, len);
            if(isValid(segment)){
                backtrack(s,start + len, dots + 1, path + segment + ".", result);
            }
        }
    }
    bool isValid(const string &segment){
        if(segment.size() > 3 || (segment[0] == '0' && segment.size() > 1)) return false;
        int val = stoi(segment);
        return val >= 0 && val <= 255;
    }
};

//iterative aaproach using nested loops
class Solution {
public:
    vector<string> restoreIpAddresses(string s) {
        vector<string> result;
        int n = s.size();
        for (int i = 1; i <= 3 && i < n - 2; ++i) {
            for (int j = i + 1; j <= i + 3 && j < n - 1; ++j) {
                for (int k = j + 1; k <= j + 3 && k < n; ++k) {
                    string s1 = s.substr(0, i);
                    string s2 = s.substr(i, j - i);
                    string s3 = s.substr(j, k - j);
                    string s4 = s.substr(k);
                    if (isValid(s1) && isValid(s2) && isValid(s3) && isValid(s4)) {
                        result.push_back(s1 + "." + s2 + "." + s3 + "." + s4);
                    }
                }
            }
        }
        return result;
    }

private:
    bool isValid(const string &segment) {
        if (segment.size() > 3 || segment.empty()) return false;
        if (segment[0] == '0' && segment.size() > 1) return false;
        int val = stoi(segment);
        return val >= 0 && val <= 255;
    }
};
